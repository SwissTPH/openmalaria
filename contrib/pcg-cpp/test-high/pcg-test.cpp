/*
 * PCG Random Number Generation for C++
 *
 * Copyright 2014-2017 Melissa O'Neill <oneill@pcg-random.org>,
 *                     and the PCG Project contributors.
 *
 * SPDX-License-Identifier: (Apache-2.0 OR MIT)
 *
 * Licensed under the Apache License, Version 2.0 (provided in
 * LICENSE-APACHE.txt and at http://www.apache.org/licenses/LICENSE-2.0)
 * or under the MIT license (provided in LICENSE-MIT.txt and at
 * http://opensource.org/licenses/MIT), at your option. This file may not
 * be copied, modified, or distributed except according to those terms.
 *
 * Distributed on an "AS IS" BASIS, WITHOUT WARRANTY OF ANY KIND, either
 * express or implied.  See your chosen license for details.
 *
 * For additional information about the PCG random number generation scheme,
 * visit http://www.pcg-random.org/.
 */

/*
 * This file is based on the demo program for the C generation schemes.
 * It shows some basic generation tasks.
 */

#include <cstddef>
#include <cstdint>
#include <cstdlib>
#include <cassert>
#include <climits>
#include <iostream>
#include <sstream>
#include <iomanip>
#include <algorithm>
#include <numeric>
#include <random>       // for random_device

#include "pcg_random.hpp"

// This code can be compiled with the preprocessor symbol RNG set to the
// PCG generator you'd like to test.

#ifndef RNG
    #define RNG pcg32
    #define TWO_ARG_INIT 1
#endif

#define STRINGIFY_IMPL(x) #x
#define STRINGIFY(x)      STRINGIFY_IMPL(x)

using namespace std;
using pcg_extras::operator<<;

#if !PCG_EMULATED_128BIT_MATH || !AWKWARD_128BIT_CODE

int main(int argc, char** argv)
{
    // Read command-line options

    int rounds = 5;
    bool nondeterministic_seed = false;

    ++argv;
    --argc;
    if (argc > 0 && strcmp(argv[0], "-r") == 0) {
        nondeterministic_seed = true;
        ++argv;
        --argc;
    }
    if (argc > 0) {
        rounds = atoi(argv[0]);
    }

    /* Many of the generators can be initialized with two arguments; the second
     * one specifies the stream.
     */

#if TWO_ARG_INIT
    RNG rng(42u, 54u);
#else
    RNG rng(42u);
#endif

    if (nondeterministic_seed) {
        // Seed with external entropy from std::random_device (a handy
        // utility provided by pcg_extras).
        rng.seed(pcg_extras::seed_seq_from<std::random_device>());
    }

    constexpr auto bits = sizeof(RNG::result_type) * CHAR_BIT;
    constexpr int how_many_nums = bits <= 8  ? 14
                                : bits <= 16 ? 10
                                :              6;
    constexpr int wrap_nums_at  = bits >  64 ? 2
                                : bits >  32 ? 3
                                :              how_many_nums;

    cout << STRINGIFY(RNG) << ":\n"
    //   << "      -  aka:         " << pcg_extras::printable_typename<RNG>()
    // ^-- we skip this line because the output is long, scary, ugly, and
    //     and varies depending on the platform
         << "      -  result:      " <<  bits << "-bit unsigned int\n"
         << "      -  period:      2^" << RNG::period_pow2();
    if (RNG::streams_pow2() > 0)
         cout << "   (* 2^" << RNG::streams_pow2() << " streams)";
    cout << "\n      -  size:        " << sizeof(RNG) << " bytes\n\n";

    for (int round = 1; round <= rounds; ++round) {
        printf("Round %d:\n", round);

        /* Make some N-bit numbers */
        cout << setw(4) << setfill(' ') << bits << "bit:";
        for (int i = 0; i < how_many_nums; ++i) {
            if (i > 0 && i % wrap_nums_at == 0)
                cout << "\n\t";
            cout << " 0x" << hex << setfill('0') 
                 << setw(sizeof(RNG::result_type)*2) << rng();
        }
        cout << endl;

        cout << "  Again:";
        rng.backstep(6);
        for (int i = 0; i < how_many_nums; ++i) {
            if (i > 0 && i % wrap_nums_at == 0)
                cout << "\n\t";
            cout << " 0x" << hex << setfill('0') 
                 << setw(sizeof(RNG::result_type)*2) << rng();
        }
        cout << dec << endl;

        /* Toss some coins */
        cout << "  Coins: ";
        for (int i = 0; i < 65; ++i)
            cout << (rng(2) ? "H" : "T");
        cout << endl;

        RNG rng_copy{rng};
        /* Roll some dice */
        printf("  Rolls:");
        for (int i = 0; i < 33; ++i)
            cout << " " << (uint32_t(rng(6)) + 1);
        cout << "\n   -->   rolling dice used "
             << static_cast<size_t>(rng - rng_copy)
             << " random numbers" << endl;

        /* Deal some cards using pcg_extras::shuffle, which follows
         * the algorithm for shuffling that most programmers would expect.
         * (It's unspecified how std::shuffle works.)
         */      
        enum { SUITS = 4, NUMBERS = 13, CARDS = 52 };
        char cards[CARDS];
        iota(begin(cards), end(cards), 0);
        pcg_extras::shuffle(begin(cards), end(cards), rng);

        /* Output the shuffled deck */
        printf("  Cards:");
        static const signed char number[] = {'A', '2', '3', '4', '5', '6', '7',
                                             '8', '9', 'T', 'J', 'Q', 'K'};
        static const signed char suit[] =   {'h', 'c', 'd', 's'};
        int i = 0;
        for (auto card : cards) {
            ++i;
            cout << " " << number[card / SUITS] << suit[card % SUITS];
            if (i % 22 == 0)
                cout << "\n\t";
        }
        
        cout << "\n" << endl;
    }

    // Ensure that the input/output functions work.
    std::stringstream ss;
    ss << rng;
    RNG from_string;
    ss >> from_string;
    assert(from_string == rng);

    return 0;
}

#else //  i.e. PCG_EMULATED_128BIT_MATH && AWKWARD_128BIT_CODE

int main()
{
    // It's not that it *can't* be done, it just requires either C++14-style
    // constexpr or some things not to be declared const.
    cout << "Sorry, " STRINGIFY(RNG) " not supported with emulated 128-bit math"
         << endl;
}

#endif
