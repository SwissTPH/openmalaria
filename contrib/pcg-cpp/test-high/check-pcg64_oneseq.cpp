#define RNG pcg64_oneseq
#define TWO_ARG_INIT 0

#include "pcg-test.cpp"

