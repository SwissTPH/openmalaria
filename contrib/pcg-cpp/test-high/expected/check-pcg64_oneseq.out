pcg64_oneseq:
      -  result:      64-bit unsigned int
      -  period:      2^128
      -  size:        16 bytes

Round 1:
  64bit: 0x287472e87ff5705a 0xbbd190b04ed0b545 0xb6cee3580db14880
	 0xbf5f7d7e4c3d1864 0x734eedbe7e50bbc5 0xa5b6b5f867691c77
  Again: 0x287472e87ff5705a 0xbbd190b04ed0b545 0xb6cee3580db14880
	 0xbf5f7d7e4c3d1864 0x734eedbe7e50bbc5 0xa5b6b5f867691c77
  Coins: HHTHHHTTHTHHTTTHHHTHHHTTTTTHHTHTTTTHHTHHTTTTTTHHTHTHTTTTTHTHHTTHT
  Rolls: 1 2 6 3 6 2 6 5 3 2 3 2 1 5 1 6 1 3 3 5 4 3 1 5 1 4 6 4 1 6 5 5 5
   -->   rolling dice used 33 random numbers
  Cards: 9c 5h 7d 7c 4c 8d 7h Qc Kh 2d 3h 2h Qd Ts 3d Kc 9h Jc 6h 6d 8c 4d
	 Qh As Jh 8s Th 5s 2c 9d Ac 4h Kd 5d 9s 6c 3s Ks Js Jd 7s 2s 3c Tc
	 Qs 4s 8h 5c Ah 6s Td Ad

Round 2:
  64bit: 0x7d97ee72fb94fdf0 0xb35f07d53cc42b66 0x0854c5caec0c251f
	 0xf37961a645554320 0x1d1d213622351b24 0x6edbb396c73fb49f
  Again: 0x7d97ee72fb94fdf0 0xb35f07d53cc42b66 0x0854c5caec0c251f
	 0xf37961a645554320 0x1d1d213622351b24 0x6edbb396c73fb49f
  Coins: HHTTHHHTHHHHHTTHHHTHHTHTHTTTHHTHHHHHHTHTTHHHHHTHTTHTHHHTHHTTHHHHH
  Rolls: 5 4 2 2 5 3 2 2 2 4 3 1 2 5 6 1 6 5 3 1 1 3 5 6 4 2 5 3 1 2 2 4 1
   -->   rolling dice used 33 random numbers
  Cards: Qh Ah 3c 6d 8s 3h Jh 8c 6s 9h 8d 3s 9c Ts Qd Kc Kh 4d 7c 5h Th 2s
	 2c 4c 6h Jd Td 2d Ks 5c 7h Qs 3d 4s 7d 5s 9d Ac 6c 9s 8h As Jc 7s
	 2h Tc Kd 4h Ad 5d Qc Js

Round 3:
  64bit: 0x187ee00430cec695 0x38efe3fb60c70613 0x3949bd01ef38c552
	 0xd3f1543a45f3b48f 0xfb81a0482dc602cd 0xb48e4f661e4c7fc5
  Again: 0x187ee00430cec695 0x38efe3fb60c70613 0x3949bd01ef38c552
	 0xd3f1543a45f3b48f 0xfb81a0482dc602cd 0xb48e4f661e4c7fc5
  Coins: THTTTHHTHHHHHHTHTTTTHHTHTHTHHHTHTTTTTHHTHTTTTTHTTTTTHHTHHHTTTTHHT
  Rolls: 1 5 6 2 1 3 2 5 5 6 3 4 4 1 5 3 5 1 4 1 2 4 4 1 3 5 6 5 4 5 5 6 3
   -->   rolling dice used 33 random numbers
  Cards: 9c 4s 9d Kh Js 6c 4d 8h Qd 8d 5s 3d 6h 7s 9s Ks 2s 7h Ac Ad Kc 4c
	 5h Jd As Ah 8s 2h Jh 6s Th 2c Qc 8c 5d 9h 3h 4h Qh 3s Tc 5c 2d Qs
	 3c Ts Kd Jc 7c 6d 7d Td

Round 4:
  64bit: 0xd04c0a3a8cf6c571 0xbc94812fe9ec2c93 0x691f3e3aa2f42c77
	 0xb7188d5162d89a1e 0x17fbf02e08fee28a 0x1aa17486e288664f
  Again: 0xd04c0a3a8cf6c571 0xbc94812fe9ec2c93 0x691f3e3aa2f42c77
	 0xb7188d5162d89a1e 0x17fbf02e08fee28a 0x1aa17486e288664f
  Coins: THTHHHTTTHHHHHTTHHHHHTTHHTTHTHTTHHHHHTHTTHTTTTHTTTHTHHHHTTHHTTHTH
  Rolls: 3 6 5 6 2 3 5 6 2 5 1 5 3 5 6 2 3 1 1 3 1 6 6 4 4 5 4 4 4 2 4 6 5
   -->   rolling dice used 33 random numbers
  Cards: 6d 4c 9c 7s Qs Qh 5d 5h Jd 8d 3c 3d 9h Th 7d Tc As 7h 9s 2d 2s Td
	 8s 4s 7c Qd Jc 5c 5s 2c 2h 6s 6c 4h Js Ks 8h 6h Jh Kh 4d 3s Ah 9d
	 3h Kc 8c Qc Kd Ts Ad Ac

Round 5:
  64bit: 0x10bda17a1292d5aa 0xf0cd1384e25b3497 0x8e592be49a6a6181
	 0x5edc4faf5cda5865 0xb2ecea43437a3f8c 0x98dbb99c3550f0e4
  Again: 0x10bda17a1292d5aa 0xf0cd1384e25b3497 0x8e592be49a6a6181
	 0x5edc4faf5cda5865 0xb2ecea43437a3f8c 0x98dbb99c3550f0e4
  Coins: HTHTHHHHHHTHHTTHHHTTTTTTHTTTHHHTHHHHHHHHHTTTTHHTTTHTHHHHTTHHHHTTH
  Rolls: 5 1 4 6 1 4 2 4 2 1 3 2 4 3 6 3 5 5 4 5 1 2 1 1 1 6 5 6 5 4 1 6 4
   -->   rolling dice used 33 random numbers
  Cards: 3d 7h Jd 3h 5d 2s 7d 5s 9d 8s 4c 8c Jc Ah Tc Js Ad 5h 6h Th 7s Kd
	 Qd 9h As 8d 3c 2d 2c Kh 9s 7c Ac 6c 9c 2h Qs 4h Ts 6s Ks Jh 5c 8h
	 3s 6d Kc 4d Td Qh Qc 4s

